ml/Database.cc

```diff
@@ -38,6 +38,80 @@ const char *ml::Database::SQL_SELECT_ANOMALY_EVENTS =
+    "  ((COALESCE(pre.avg,0)/(CASE WHEN COALESCE(pre.avg,1) = 0 THEN 0.0000001 ELSE COALESCE(pre.avg,1) END) * "
+    "   (SELECT before-?2 FROM anomaly_rate_info WHERE after <= ?2 ORDER BY after DESC LIMIT 1)) + "
+    "  (COALESCE(main.avg,0)/(CASE WHEN COALESCE(main.avg,1) = 0 THEN 0.0000001 ELSE COALESCE(main.avg,1) END) * "
+    "   ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) - "
+    "    (SELECT after FROM anomaly_rate_info WHERE after >= ?2 ORDER BY after ASC LIMIT 1))) + "
+    "  (COALESCE(post.avg,0)/(CASE WHEN COALESCE(post.avg,1) = 0 THEN 0.0000001 ELSE COALESCE(post.avg,1) END) * "
+    "   (SELECT ?3-after FROM anomaly_rate_info WHERE before >= ?3 OR "
+    "                 ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) "
+    "                  AND NOT EXISTS (SELECT before FROM anomaly_rate_info WHERE before >= ?3 ORDER BY before ASC LIMIT 1)) "
+    "               ORDER BY before ASC LIMIT 1))) "
+    " ),0.0) percentage FROM "
+    "(SELECT dimension_id, AVG(anomaly_percentage) avg FROM "
+    "  (SELECT json_extract(j.value, '$[1]') AS dimension_id, "
+    "  json_extract(j.value, '$[0]') AS anomaly_percentage "
+    "  FROM anomaly_rate_info AS ari, json_each(ari.anomaly_rates) AS j " 
+    "  WHERE ari.host_id == ?1 AND ari.after >= ?2 AND ari.before <= ?3 "
+    "  AND json_valid(ari.anomaly_rates)) "
+    "  GROUP BY dimension_id) AS main "
+    " LEFT JOIN "
+    "(SELECT dimension_id, AVG(anomaly_percentage) avg FROM "
+    "  (SELECT json_extract(j.value, '$[1]') AS dimension_id, "
+    "  json_extract(j.value, '$[0]') AS anomaly_percentage "
+    "  FROM anomaly_rate_info AS ari, json_each(ari.anomaly_rates) AS j " 
+    "  WHERE ari.host_id == ?1 "
+    "  AND ari.after >= (SELECT after FROM anomaly_rate_info WHERE after <= ?2 ORDER BY after DESC LIMIT 1) "
+    "  AND ari.before <= (SELECT before FROM anomaly_rate_info WHERE before >= ?2 ORDER BY before ASC LIMIT 1) "
+    "  AND json_valid(ari.anomaly_rates)) "
+    "  GROUP BY dimension_id) AS pre "
+    " ON main.dimension_id = pre.dimension_id "
+    " LEFT JOIN "
+    "(SELECT dimension_id, AVG(anomaly_percentage) avg FROM "
+    "  (SELECT json_extract(j.value, '$[1]') AS dimension_id, "
+    "  json_extract(j.value, '$[0]') AS anomaly_percentage "
+    "  FROM anomaly_rate_info AS ari, json_each(ari.anomaly_rates) AS j " 
+    "  WHERE ari.host_id == ?1 "
+    "  AND ari.after >= (SELECT after FROM anomaly_rate_info WHERE after <= ?3 ORDER BY after DESC LIMIT 1) "
+    "  AND ari.before <= (SELECT before FROM anomaly_rate_info WHERE before >= ?3 OR "
+    "                ((SELECT before FROM anomaly_rate_info WHERE before <= ?3 ORDER BY before DESC LIMIT 1) "
+    "                 AND NOT EXISTS (SELECT before FROM anomaly_rate_info WHERE before >= ?3 ORDER BY before ASC LIMIT 1)) "
+    "              ORDER BY before DESC LIMIT 1) "
```

Comment:
Hi Siamak!

I'm trying to understand this part of the sql query.

I'm assuming that the goal here from line 93 -> 104, is to get the very next row in the database after the value of `before`.

So if we have a sample db:

|after|before|
|------|---------|
|10|20|
|20|30|
|30|40|
|40|50|

And we request after = 10, before = 35, then this select wants also to include part of 30 -> 40, but not anything from 40 -> 50, right?

If indeed in the sql query `?2` is after, and `?3` is before, then it appears that it will include also values from 40 -> 50 and everything too after that if they exist. Is this expected ? Does the rest of the query somehow handles that?

Can't `ari.before` be simply `SELECT before FROM anomaly_rate_info_test WHERE before >= ?3 ORDER BY before ASC LIMIT 1` ? Can you help me understand the logic of how `ari.before` is set?

Thanks!

Reply:
Thanks Manolis,
The 'post' part of the query has to be cautious in case there is no next saved record after the given before parameter (?3). If you choose to have only the sub-query SELECT before FROM anomaly_rate_info_test WHERE before >= ?3 ORDER BY before ASC LIMIT 1, then what if it does not return any? it has to bring the closest or the latest saved before timestamp.

Reply:
If DESC in line 102 is replaced by ASC, as it was before, it brings the immediate next before record. I changed it to DESC in response to another issue, need to check if it affects any other case!